nav2_fake_localization/src/fake_localization.cpp

```diff
@@ -0,0 +1,176 @@
+
+  initialpose_sub_.subscribe(this, "initialpose");
+  initialpose_filter_ = std::make_shared<
+    tf2_ros::MessageFilter<geometry_msgs::msg::PoseWithCovarianceStamped>>(
+    initialpose_sub_, *tf_buffer_, global_frame_id_, 1,
+    this->get_node_logging_interface(), this->get_node_clock_interface(),
+    buffer_timeout);
+  initialpose_filter_->registerCallback(&FakeOdomNode::sub_initialpose, this);
+}
+
+FakeOdomNode::~FakeOdomNode() {}
+
+void FakeOdomNode::sub_odom(const nav_msgs::msg::Odometry::SharedPtr msg)
+{
+  tf2::Transform txi;
+  tf2::convert(msg->pose.pose, txi);
+  txi = tf_offset_ * txi;
+
+  geometry_msgs::msg::TransformStamped odom_to_map;
+  try {
+    geometry_msgs::msg::TransformStamped txi_inv;
+    txi_inv.header.frame_id = base_frame_id_;
+    txi_inv.header.stamp = msg->header.stamp;
+    tf2::convert(txi.inverse(), txi_inv.transform);
+
+    tf_buffer_->transform(txi_inv, odom_to_map, odom_frame_id_);
+  } catch (tf2::TransformException & e) {
+    RCLCPP_ERROR(
+      get_logger(), "Failed to transform to %s from %s: %s\n",
+      odom_frame_id_.c_str(), base_frame_id_.c_str(), e.what());
+    return;
+  }
+
+  geometry_msgs::msg::TransformStamped trans;
+
+  trans.header.stamp = msg->header.stamp;
+  trans.header.stamp.nanosec += 1e8;
+  trans.header.frame_id = global_frame_id_;
+  trans.child_frame_id = msg->header.frame_id;
+  tf2::Transform odom_to_map_tf2;
+  tf2::convert(odom_to_map.transform, odom_to_map_tf2);
+  tf2::Transform odom_to_map_inv = odom_to_map_tf2.inverse();
+  tf2::convert(odom_to_map_inv, trans.transform);
+  tf_broadcaster_->sendTransform(trans);
+}
+
+void FakeOdomNode::sub_initialpose(
+  const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
+{
+  tf2::Transform pose;
+  tf2::convert(msg->pose.pose, pose);
+
+  if (msg->header.frame_id != global_frame_id_) {
+    RCLCPP_WARN(
+      get_logger(),
+      "Frame ID of \"initialpose\" (%s) is different from the global "
+      "frame %s",
+      msg->header.frame_id.c_str(), global_frame_id_.c_str());
+  }
+
+  // set offset so that current pose is set to "initialpose"
+  geometry_msgs::msg::TransformStamped base_in_map;
+  try {
+    // just get the latest
+    base_in_map = tf_buffer_->lookupTransform(
+      base_frame_id_, global_frame_id_,
+      rclcpp::Time(0));
+  } catch (tf2::TransformException & e) {
+    RCLCPP_WARN(get_logger(), "Failed to lookup transform! %s", e.what());
+    return;
+  }
+
+  tf2::Transform base_in_map2;
+  tf2::convert(base_in_map.transform, base_in_map2);
+  tf2::Transform delta = pose * base_in_map2;
+  tf_offset_ = delta * tf_offset_;
+}
+
+} // namespace nav2_fake_localization
+
+int main(int argc, char * argv[])
+{
+  rclcpp::init(argc, argv);
+  auto node = std::make_shared<nav2_fake_localization::FakeOdomNode>();
+  rclcpp::Rate loop(20);
+
+  while (rclcpp::ok()) {
+    rclcpp::spin_some(node);
+    loop.sleep();
+  }
+  rclcpp::shutdown();
+  return 0;
+}
```

Comment:
Remove main, make this a component. You can have a `main` in another file, but that way it can be used as a component or as a standalone node